File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int dfs(TreeNode* root){
                if(root == NULL){
            return 0;
        }

        int left = 0;
        int right = 0;


        if(root->left){
            left = dfs(root->left);
        }

        if(root->right){
            right = dfs(root->right);
        }
        res = max(left+right, res);

        return max(left, right) + 1;

    }
    int diameterOfBinaryTree(TreeNode* root) {
        
        dfs(root);
        return res;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def getDiameter(self,root): 

        if root == None:
            return [0,0]
        
        [lDiameter, lheight] = self.getDiameter(root.left)

        [rDiameter, rheight] = self.getDiameter(root.right)

        diameter = lheight + rheight

        return [max(diameter, lDiameter, rDiameter), max(lheight, rheight) + 1] 

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        [res, height] = self.getDiameter(root)
        return res

                





        